src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs

```diff
@@ -0,0 +1,895 @@
+    // (24,31): error CS0190: The __arglist construct is valid only within a variable argument method
+    //             Console.WriteLine(__arglist);
+    Diagnostic(ErrorCode.ERR_ArgsInvalid, "__arglist").WithLocation(24, 31),
+    // (31,31): error CS4013: Instance of type 'RuntimeArgumentHandle' cannot be used inside an anonymous function, query expression, iterator block or async method
+    //             Console.WriteLine(__arglist);
+    Diagnostic(ErrorCode.ERR_SpecialByRefInLambda, "__arglist").WithArguments("System.RuntimeArgumentHandle").WithLocation(31, 31)
+                );
+        }
+
+        [Fact]
+        public void ByRefIterator()
+        {
+            var source = @"
+using System;
+using System.Collections.Generic;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        IEnumerable<int> Local(ref int x)
+        {
+            yield return x;
+        }
+        int y = 2;
+        Console.WriteLine(string.Join("","", Local(ref y)));
+    }
+}
+";
+            var option = TestOptions.ReleaseExe.WithWarningLevel(0);
+            CreateCompilationWithMscorlibAndSystemCore(source, options: option, parseOptions: _parseOptions).VerifyDiagnostics(
+    // (9,40): error CS1623: Iterators cannot have ref or out parameters
+    //         IEnumerable<int> Local(ref int x)
+    Diagnostic(ErrorCode.ERR_BadIteratorArgType, "x").WithLocation(9, 40)
+                );
+        }
+
+        [Fact]
+        public void Extension()
+        {
+            var source = @"
+using System;
+using System.Collections.Generic;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        int Local(this int x)
+        {
+            return x;
+        }
+        Console.WriteLine(Local(2));
+    }
+}
+";
+            var option = TestOptions.ReleaseExe.WithWarningLevel(0);
+            CreateCompilationWithMscorlibAndSystemCore(source, options: option, parseOptions: _parseOptions).VerifyDiagnostics(
+    // (9,13): error CS1106: Extension method must be defined in a non-generic static class
+    //         int Local(this int x)
+    Diagnostic(ErrorCode.ERR_BadExtensionAgg, "Local").WithLocation(9, 13)
+                );
+        }
+
+        [Fact]
+        public void InferredReturn()
+        {
+            var source = @"
+using System;
+using System.Collections.Generic;
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        var Local()
+        {
+            return 2;
+        }
+        Console.WriteLine(Local());
+    }
+}
+";
+            // message is temporary
+            var option = TestOptions.ReleaseExe.WithWarningLevel(0);
+            CreateCompilationWithMscorlibAndSystemCore(source, options: option, parseOptions: _parseOptions).VerifyDiagnostics(
+    // (9,9): error CS7019: Type of 'Program.Local()' cannot be inferred since its initializer directly or indirectly refers to the definition.
```

Comment:
I figured one of two things would happen: inferred return types would be allowed, in which case this message would be accurate (the above case would compile file without diagnostics, but other cases would not due to the messages reason), or we disallow inferred return types, in which case another message can be added. I tended towards not adding/modifying error IDs or messages for now to keep things simple.